tests/e2e/hashkv_test.go

```diff
@@ -0,0 +1,359 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !cluster_proxy
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/etcd/client/pkg/v3/fileutil"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	fe2e "go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+func TestVerifyHashKVAfterCompact(t *testing.T) {
+	tests := []struct {
+		compactedOnTombstoneRev bool
+		hashKVOnTombstoneRev    bool
+	}{
+		{
+			compactedOnTombstoneRev: true,
+			hashKVOnTombstoneRev:    true,
+		},
+		{
+			compactedOnTombstoneRev: false,
+			hashKVOnTombstoneRev:    true,
+		},
+		{
+			compactedOnTombstoneRev: true,
+			hashKVOnTombstoneRev:    false,
+		},
+		{
+			compactedOnTombstoneRev: false,
+			hashKVOnTombstoneRev:    false,
+		},
+	}
+
+	scenarios := []struct {
+		ClusterVersion fe2e.ClusterVersion
+		OnlyOneKey     bool
+	}{
+		{
+			ClusterVersion: fe2e.CurrentVersion,
+			OnlyOneKey:     true,
+		},
+		{
+			ClusterVersion: fe2e.CurrentVersion,
+			OnlyOneKey:     false,
+		},
+		{
+			ClusterVersion: fe2e.QuorumLastVersion,
+			OnlyOneKey:     true,
+		},
+		{
+			ClusterVersion: fe2e.QuorumLastVersion,
+			OnlyOneKey:     false,
+		},
+	}
+
+	for idx, tt := range tests {
+		for _, scenario := range scenarios {
+			t.Run(fmt.Sprintf("#%d - %s - OnlyOneKey=%v", idx, scenario.ClusterVersion, scenario.OnlyOneKey), func(t *testing.T) {
+				ctx := context.Background()
+
+				clus, cliCfg := newClusterForHashKV(t, scenario.ClusterVersion)
+
+				dataset := newTestDatasetInCluster(t, clus, cliCfg, scenario.OnlyOneKey)
+
+				// ensure we can pick up the right revision
+				require.True(t, len(dataset.tombstones) > 2)
+				require.True(t, dataset.tombstones[0]+3 < dataset.tombstones[1])
+				require.True(t, dataset.tombstones[1]+3 < dataset.tombstones[2])
+
+				compactedOnRev := dataset.tombstones[0]
+				if !tt.compactedOnTombstoneRev {
+					compactedOnRev += 3
+				}
+
+				hashKVOnRev := dataset.tombstones[1]
+				if !tt.hashKVOnTombstoneRev {
+					hashKVOnRev += 3
+				}
+
+				cli, err := fe2e.NewEtcdctl(cliCfg, clus.EndpointsGRPC())
+				require.NoError(t, err)
+
+				t.Logf("COMPACT on rev=%d", compactedOnRev)
+				_, err = cli.Compact(ctx, compactedOnRev, config.CompactOption{Physical: true})
+				require.NoError(t, err)
+
+				t.Logf("HashKV on rev=%d", hashKVOnRev)
+				resp, err := cli.HashKV(ctx, hashKVOnRev)
+				require.NoError(t, err)
+
+				require.Len(t, resp, 3)
+				require.True(t, resp[0].Hash != 0)
+				t.Logf("One Hash value is %d", resp[0].Hash)
+
+				require.Equal(t, resp[0].Hash, resp[1].Hash)
+				require.Equal(t, resp[1].Hash, resp[2].Hash)
+			})
+		}
+	}
+}
+
+type datasetInfo struct {
+	keys map[string]struct{}
+
+	tombstones      []int64
+	lastestRevision int64
+}
+
+// key: "foo"
```

Comment:
Thanks for crafting such a meticulous testing scenario, this seems very throughout, however for reviewer it's or any future contributor it might be hard to understand why we picked this exact data shape and no other. 

My question would be, how hard would it be to generalize? We know one an exact case that we know that we could break hashKV, but how sure we are that we don't miss another case? Hardcoding a exact scenario protects us against this exact mistake, but can we be sure there will not be more such cases in the future? Can we be sure that HashKV, is resilient enough?

Have you heard about property testing? https://www.mayhem.security/blog/what-is-property-based-testing. For me the etcd should always hold property that no matter the type of write, no matter the compaction revision, no matter the etcd version, no matter the revision, HashKV should always be the same on all cluster members.

I think it would be ok to merge those exact scenarios, but I also we should think how we can generalize it. For example, instead of checking hashKV on one revision, test all of them in for loop.

Reply:
> instead of checking hashKV on one revision, test all of them in for loop.

Good point. Will simplify data generator.

Reply:
I follow @ahrtr suggestion with using `populateData`. Please take a look.